priam/src/main/java/com/netflix/priam/PriamServer.java

```diff
@@ -111,25 +115,48 @@ public void scheduleService() throws Exception {
                     UpdateSecuritySettings.getTimer(instanceIdentity));
         }
 
-        // Set up cassandra tuning.
-        cassandraTunerService.scheduleService();
+        // Set up the background configuration dumping thread
+        scheduleTask(
+                scheduler,
+                PriamConfigurationPersister.class,
+                PriamConfigurationPersister.getTimer(config));
+
+        boolean shouldStartCassandra = false;
 
-        // Determine if we need to restore from backup else start cassandra.
-        if (restoreContext.isRestoreEnabled()) {
+        // Determine if we need to restore from backup.
+        if (restoreContext.isRestoreEnabled(config, instanceIdentity.getInstanceInfo())) {
             restoreContext.restore();
-        } else { // no restores needed
-            logger.info("No restore needed, task not scheduled");
-            if (!config.doesCassandraStartManually()) cassProcess.start(true); // Start cassandra.
-            else
-                logger.info(
-                        "config.doesCassandraStartManually() is set to True, hence Cassandra needs to be started manually ...");
+            // Start cassandra only if restore is successful.
+            shouldStartCassandra = true;
+        } else {
+            if (instanceIdentity.isReplace()
+                    && backupRestoreConfig.enableBypassCassandraStreaming()) {
+                logger.info("Trying to download data instead of streaming from Cassandra.");
+                try {
+                    restoreContext.restore();
+                    instanceIdentity.setReplacedIp("");
+                } catch (Exception e) {
+                    logger.error(
+                            "Error while trying to rebuild the node from backup. Maybe backup not available or disk full? Trying normal path of cassandra streaming");
+                    // Clean the data folder.
+                    SystemUtils.cleanupDir(config.getDataFileLocation(), null);
+                } finally {
+                    shouldStartCassandra = true;
+                }
+            } else {
+                // no restores needed
+                logger.info("No restore needed, task not scheduled");
+                shouldStartCassandra = true;
+            }
         }
```

Comment:
How can we exit this block with shouldStartCassandra being false?

Reply:
If the original requested `restore` was a failure (that is when Priam starts in restore mode during weekly restore refresh). `
`` `       // Start cassandra only if restore is successful.
            shouldStartCassandra = true;```

But yes with the recent refactoring of restore we will throw exception there and thus we don't need that variable. good catch.